File: arithmetic_progression.py
def is_arithmetic_progression(nums):
    nums.sort()
    return len(set(b - a for a, b in zip(nums, nums[1:]))) == 1


def main():
    nums1 = [1, 3, 5, 7, 9]
    nums2 = [9, 1, 7, 5, 3]
    nums3 = [1, 2, 4, 8, 16]
    nums4 = [5, -1, 3, 1, -3]
    nums5 = [1.5, 3.0, 0.0, 4.5, 6.0]

    print(f"Example 1:\nInput: {nums1}\nOutput: {is_arithmetic_progression(nums1)}\n")
    print(f"Example 2:\nInput: {nums2}\nOutput: {is_arithmetic_progression(nums2)}\n")
    print(f"Example 3:\nInput: {nums3}\nOutput: {is_arithmetic_progression(nums3)}\n")
    print(f"Example 4:\nInput: {nums4}\nOutput: {is_arithmetic_progression(nums4)}\n")
    print(f"Example 5:\nInput: {nums5}\nOutput: {is_arithmetic_progression(nums5)}\n")


if __name__ == "__main__":
    main()

File: arithmetic_progression.cpp
// https://godbolt.org/z/eeofxrboW

#include <algorithm>
#include <functional>
#include <print>
#include <ranges>
#include <vector>

auto is_arithmetic_progression(auto nums) -> bool {
    std::ranges::sort(nums);
    auto diffs = nums | std::views::adjacent_transform<2>(std::minus{});
    return std::ranges::adjacent_find(diffs, std::not_equal_to{}) == diffs.end();
}

int main() {
    auto nums1 = std::vector{1, 3, 5, 7, 9};
    auto nums2 = std::vector{9, 1, 7, 5, 3};
    auto nums3 = std::vector{1, 2, 4, 8, 16};
    auto nums4 = std::vector{5, -1, 3, 1, -3};
    auto nums5 = std::vector{1.5, 3., 0., 4.5, 6.};

    std::println("Example 1:\nInput: {}\nOutput: {}\n", nums1, is_arithmetic_progression(nums1));
    std::println("Example 2:\nInput: {}\nOutput: {}\n", nums2, is_arithmetic_progression(nums2));
    std::println("Example 3:\nInput: {}\nOutput: {}\n", nums3, is_arithmetic_progression(nums3));
    std::println("Example 4:\nInput: {}\nOutput: {}\n", nums4, is_arithmetic_progression(nums4));
    std::println("Example 5:\nInput: {}\nOutput: {}\n", nums5, is_arithmetic_progression(nums5));

    return 0;
}
